gems/aws-sdk-dynamodbstreams/lib/aws-sdk-dynamodbstreams/attribute_translator.rb

```diff
@@ -0,0 +1,40 @@
+module Aws
+  module DynamoDBStreams
+    # A utility class that translates DynamoDBStream events from Lambda
+    # functions into their simple attribute equivalents.
+    class AttributeTranslator
+      # Parse a DynamoDBStream event hash from Lambda that contains 1 or more
+      # records. When using the SDK to retrieve DynamoDB stream records, use the
+      # `simple_attributes: true` option instead.
+      #
+      # @param [Hash] event A DynamoDBStream event.
+      #
+      # @example Parse a DynamoDB stream event from Lambda
+      #   def lambda_handler(event:, context:)
+      #     records = Aws::DynamoDBStreams::AttributeTranslator.from_event(event)
+      #     records.each do |record|
+      #       puts record.dynamodb.new_image
+      #       # => { "size" => 123, "enabled" => true, ... }
+      #     end
+      #   end
+      def self.from_event(event)
+        return unless event.is_a?(Hash)
+
+        # TODO: This implementation is slow and inefficient. It would be
+        # better to write a switch-case that has similar logic to AttributeValue
+        # used in the ValueTranslator. This implementation however provides
+        # consistency between both DynamoDB and DynamoDBStreams. The translator
+        # only works with shapes and structs and not a regular Ruby hash.
+        shape_ref = ClientApi::Shapes::ShapeRef.new(
+          shape: ClientApi::GetRecordsOutput
+        )
+        translator = Plugins::SimpleAttributes::ValueTranslator.new(
+          shape_ref, :unmarshal
+        )
+        parser = Aws::Json::Parser.new(shape_ref)
+        input = parser.parse(JSON.dump(event))
```

Comment:
If you wanted to drop re-encoding the event to JSON, maybe remove the private qualifier here

https://github.com/aws/aws-sdk-ruby/blob/203aaba7242357884fe182be8f2d36f77428b8a7/gems/aws-sdk-core/lib/aws-sdk-core/json/parser.rb#L22

or copy this from the JSON parser:

https://github.com/aws/aws-sdk-ruby/blob/203aaba7242357884fe182be8f2d36f77428b8a7/gems/aws-sdk-core/lib/aws-sdk-core/json/parser.rb#L24-L72

Calling `parse_ref` on the event Hash looks like it'll resolve the Hash -> JSON -> Struct sequence to just Hash -> Struct.

Quick and dirty demo on the existing sdk's ddb item

```ruby
module Demo
  include Seahorse::Model::Shapes
  
  module_function

  def translate(image)
    shape_ref = Aws::DynamoDB::ClientApi::Shapes::ShapeRef.new(shape: Aws::DynamoDB::ClientApi::GetItemOutput)
    translator = Aws::DynamoDB::Plugins::SimpleAttributes::ValueTranslator.new(shape_ref, :unmarshal)
    translator.apply(parse_ref(shape_ref, { 'Item' => image })).item
  end

  def parse_ref(ref, value, target = nil)
    if value.nil?
      nil
    else
      case ref.shape
      when StructureShape then structure(ref, value, target)
      when ListShape then list(ref, value, target)
      when MapShape then map(ref, value, target)
      when TimestampShape then time(value)
      when BlobShape then Base64.decode64(value)
      when BooleanShape then value.to_s == 'true'
      else value
      end
    end
  end

  def structure(ref, values, target = nil)
    shape = ref.shape
    target = ref.shape.struct_class.new if target.nil?
    values.each do |key, value|
      member_name, member_ref = shape.member_by_location_name(key)
      if member_ref
        target[member_name] = parse_ref(member_ref, value)
      end
    end
    target
  end

  def list(ref, values, target = nil)
    target = [] if target.nil?
    values.each do |value|
      target << parse_ref(ref.shape.member, value)
    end
    target
  end

  def map(ref, values, target = nil)
    target = {} if target.nil?
    values.each do |key, value|
      target[key] = parse_ref(ref.shape.value, value)
    end
    target
  end

  def time(value)
    value.is_a?(Numeric) ? Time.at(value) : Time.parse(value)
  end
end
```

Using part of Lambda's ddbEvent test input

```ruby
[10] pry(main)> ddb_event['Records'].first['dynamodb']['NewImage']
=> {"Message"=>{"S"=>"New item!"}, "Id"=>{"N"=>"101"}}
[11] pry(main)> Demo.translate(ddb_event['Records'].first['dynamodb']['NewImage'])
=> {"Message"=>"New item!", "Id"=>0.101e3}
[12] pry(main)>
```

Reply:
Thanks for the suggestion. The re-encoding is not actually the slowest part.. it's mainly because of the shape traversal. Some quick benchmarking shows the re-encoding to be about 1/4 or 1/5th the overall speed. Also "slow" is a bit of an exaggeration, it's still plenty fine. I think the best thing to do would be to make a switch case similar to that of the AttributeValue unmarshaler and guarantee that it's compatible. For now I am going to ship this.